raft/src/main/java/org/apache/kafka/raft/LeaderState.java

```diff
@@ -107,53 +112,106 @@ private OptionalLong quorumMajorityFetchTimestamp() {
 
     /**
      * @return The updated lower bound of fetch timestamps for a majority of quorum; -1 indicating that we have
-     *         not received fetch from the majority yet
+     *         not received fetch from the majority yet, or the fetch is from an observer
      */
     public OptionalLong updateFetchTimestamp(int nodeId, long timestamp) {
-        ReplicaState state = ensureValidVoter(nodeId);
-        // To be resilient to system time shifts we do not strictly require the timestamp be monotonically increasing
-        state.lastFetchTimestamp = OptionalLong.of(Math.max(state.lastFetchTimestamp.orElse(-1L), timestamp));
-        return quorumMajorityFetchTimestamp();
+        ReplicaState state = getReplicaState(nodeId);
```

Comment:
This is a meta comment: today we have two call traces to update the fetch timeout / replica end offset:

1) when local (the leader) appends, we call `updateLocalEndOffset` and then `maybeUpdateFetchTimerWithRemoteFetchTimestamp` with localId.

2) when received fetch, we call `state.updateEndOffset` and then `maybeUpdateFetchTimerWithRemoteFetchTimestamp` with remoteId.

Basically, the key is that whenever we are updating the replica state, we should always try to update both the fetch timestamp and the end offset, but today they are called from two traces, and has the risk to diverge (e.g. we may only update the timestamp but forget the end offset).

I'd suggest we consolidate these two functions to sth like `updateReplicaState(nodeId, fetchTimestamp, logEndOffset)`, so that we would never do a partial update.

Reply:
Sg, I have thought about that as well.